### PR TITLE
DAOS-10922 control: Implement ipmctl for aarch64 (#10922)

### PR DESCRIPTION
**Pull request by fengchunsong**
Implement NvmMgmt and IpmCtl interface for aarch64 without link libipmctl.

Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>

### CODE REVIEW

**Comment by fengchunsong** [comment]
@tanabarr @mjmac please help review this patch,thanks.

**Review by tanabarr — CHANGES_REQUESTED** [review]
I think the approach should be to have a "nvm.go" file which just contains the interface and the generic definitions/functions, then the struct and methods that implement the interface should be exposed in architecture specific files e.g. nvm_arm64.go and nvm_defaultarch.go (which would then include the cgo specifics). The built constraints can simply be specified as "//go:build (linux && arm64)" in nvm_arm64.go and "//go:build (!linux || !arm64)" in nvm_defaultarch.go. The "// ++build..." line is then unnecessary.

File: src/control/lib/ipmctl/nvm_ctest.go [hunk 1]
```diff
@@ -7,7 +7,6 @@
 package ipmctl
```
Comment (tanabarr) [inline src/control/lib/ipmctl/nvm_ctest.go:8]: This file should only be built if nvm_defaultarch.go is also being built, so should have the same build constraints.

Comment (fengchunsong) [reply src/control/lib/ipmctl/nvm_ctest.go:8]: Ok, Thanks.

**Comment by fengchunsong** [comment]
@tanabarr please review again. The "// ++build..." line is then unnecessary. I try remove +build, it report build error:

**Comment by fengchunsong** [comment]
@tanabarr Can you review the code again? I am new to go language, implementation may not be professional, if you have a better implementation, please help me correct.

**Review by tanabarr — APPROVED** [review]
yes this looks good to me, thank you

**Comment by tanabarr** [comment]
> @tanabarr please review again. The "// ++build..." line is then unnecessary. I try remove +build, it report build error:server/storage/scm/ipmctl.go:20:2: //go:build comment without // +build comment

yes I seem to have been wrong about this, I was going from the documentation

**Review by mjmac — COMMENTED** [review]
Hmm. Sorry if this is conflicting guidance, but I think it would probably be best in the long run to organize the code more conservatively.

It seems a reasonable assumption to me that this package will only ever be supported and functional on linux_amd64, so we should probably rename `nvm_defaultarch.go` -> `nvm_linux_amd64.go`. You can then remove the build tags from the top of the file, as the GOOS/GOARCH suffixes accomplish the same goal. Likewise, `nvm_ctest.go` could be renamed `nvm_ctest_linux_amd64.go`. Finally, I recommend renaming `nvm_arm64.go` -> `nvm_stubs.go` and using the `!(linux && amd64)` build tags so that the compiler will choose the stubs for all other platforms where this might be compiled.

The only other concern I have with this approach is that I wonder if the stub functions should really be returning `nil` errors instead of a sentinel error like `ErrUnsupportedPlatform`. I think I'm fine with addressing that later, though, as it would require more invasive changes to the rest of the code to handle that error, and it occurs to me that we could use that to solve some other problems we have in certain environments where ipmctl is not available.

Have you been able to test that the control plane operates correctly on arm64 with the stubs in place?

**Comment by fengchunsong** [comment]
@mjmac I have tested that the control plane operates correctly on arm64 with the stubs in place.ARM64 returns NIL using ipmctlstub, metrics can use Prometheus + Grafana monitoring normally.

**Comment by fengchunsong** [comment]
I also agree with you, ipmctl is an Optane tool. There is only AMD64 version, and other architectures are not supported.

**Comment by fengchunsong** [comment]
@mjmac The stub functions should really be returning nil errors instead of a sentinel error like ErrUnsupportedPlatform.  
This modification involves related modules and is not modified in this patch.

**Review by tanabarr — DISMISSED** [review]

**Review by mjmac — DISMISSED** [review]
LGTM

**Comment by johannlombardi** [comment]
@fengchunsong Could you please run src/control/run_go_tests.sh? They are failing on your PR according to internal CI.

[2022-07-02T08:52:46.640Z] ERROR: Your code hasn't been run through gofmt!
[2022-07-02T08:52:46.640Z] Please configure your editor to run gofmt on save.
[2022-07-02T08:52:46.640Z] Alternatively, at a minimum, run the following command:
[2022-07-02T08:52:46.640Z] find /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/ -name '*.go' -and -not -path '*vendor*'| xargs gofmt -w
[2022-07-02T08:52:46.640Z] 
[2022-07-02T08:52:46.640Z] gofmt check found the following:
[2022-07-02T08:52:46.640Z] 
[2022-07-02T08:52:46.640Z] diff -u /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm.go.orig /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm.go
[2022-07-02T08:52:46.640Z] --- /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm.go.orig	2022-07-02 08:52:45.953256314 +0000
[2022-07-02T08:52:46.640Z] +++ /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm.go	2022-07-02 08:52:45.953256314 +0000
[2022-07-02T08:52:46.640Z] @@ -10,7 +10,6 @@
[2022-07-02T08:52:46.640Z]  // CGO_CFLAGS & CGO_LDFLAGS env vars can be used
[2022-07-02T08:52:46.640Z]  // to specify additional dirs.
[2022-07-02T08:52:46.640Z]  
[2022-07-02T08:52:46.640Z] -
[2022-07-02T08:52:46.640Z]  /*
[2022-07-02T08:52:46.640Z]  
[2022-07-02T08:52:46.640Z]  #include "nvm_management.h"
[2022-07-02T08:52:46.640Z] @@ -43,4 +42,3 @@
[2022-07-02T08:52:46.640Z]  		initialized bool
[2022-07-02T08:52:46.640Z]  	}
[2022-07-02T08:52:46.640Z]  )
[2022-07-02T08:52:46.640Z] -
[2022-07-02T08:52:46.640Z] diff -u /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm_stubs.go.orig /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm_stubs.go
[2022-07-02T08:52:46.640Z] --- /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm_stubs.go.orig	2022-07-02 08:52:45.950256327 +0000
[2022-07-02T08:52:46.640Z] +++ /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-9596@4/src/control/lib/ipmctl/nvm_stubs.go	2022-07-02 08:52:45.950256327 +0000
[2022-07-02T08:52:46.640Z] @@ -3,8 +3,8 @@
[2022-07-02T08:52:46.640Z]  //
[2022-07-02T08:52:46.640Z]  // SPDX-License-Identifier: BSD-2-Clause-Patent
[2022-07-02T08:52:46.640Z]  //
[2022-07-02T08:52:46.640Z] -//go:build (!linux || !amd64)
[2022-07-02T08:52:46.640Z] -// +build !linux linux,!amd64
[2022-07-02T08:52:46.640Z] +//go:build !linux || !amd64
[2022-07-02T08:52:46.640Z] +// +build !linux !amd64
[2022-07-02T08:52:46.640Z]  
[2022-07-02T08:52:46.640Z]  // Package ipmctl provides Go bindings for libipmctl Native Management API
[2022-07-02T08:52:46.640Z]  package ipmctl
[2022-07-02T08:52:46.640Z] @@ -27,7 +27,6 @@
[2022-07-02T08:52:46.640Z]  	return nil
[2022-07-02T08:52:46.640Z]  }
[2022-07-02T08:52:46.640Z]  
[2022-07-02T08:52:46.640Z] -
[2022-07-02T08:52:46.640Z]  func getDevices(log logging.Logger, devs *C.struct_device_discovery, count C.NVM_UINT8) C.int {
[2022-07-02T08:52:46.640Z]  	return 0
[2022-07-02T08:52:46.640Z]  }
[2022-07-02T08:52:46.640Z] 
[2022-07-02T08:52:46.640Z] 
[2022-07-02T08:52:46.640Z] real	0m0.711s
[2022-07-02T08:52:46.640Z] user	0m1.188s
[2022-07-02T08:52:46.640Z] sys	0m0.245s
[2022-07-02T08:52:46.640Z] Test src/control/run_go_tests.sh failed with exit status 1.

**Review by johannlombardi — CHANGES_REQUESTED** [review]
Please address gofmt errors reported via the go unit test.

**Review by johannlombardi — DISMISSED** [review]

**Pull request merged**